src/Support/ExceptionHandler.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+namespace Spatie\Ray\Support;
+
+use Spatie\Ray\Ray;
+
+class ExceptionHandler
+{
+    public function catch(Ray $ray, $callback): Ray
+    {
+        $this->processCallback($ray, $callback);
+
+        if (! empty(Ray::$caughtExceptions)) {
+            throw array_shift(Ray::$caughtExceptions);
+        }
+
+        return $ray;
+    }
+
+    protected function handleCallable(Ray $ray, $callback, $rethrow = true): Ray
+    {
+        $paramType = $this->getParamType(new \ReflectionFunction($callback));
+        $expectedClasses = $this->getExpectedClasses($paramType);
+
+        if (count($expectedClasses)) {
+            $isExpected = false;
+
+            foreach ($expectedClasses as $expectedClass) {
+                foreach (Ray::$caughtExceptions as $caughtException) {
+                    if (is_a($caughtException, $expectedClass, true)) {
+                        $isExpected = true;
+
+                        break 2;
+                    }
+                }
+            }
+
+            if (! $isExpected && ! $rethrow) {
+                return $ray;
+            }
+
+            if (! $isExpected && $rethrow) {
+                throw array_shift(Ray::$caughtExceptions);
+            }
+        }
+
+        $exception = array_shift(Ray::$caughtExceptions);
+
+        $callbackResult = $callback($exception, $ray);
+
+        return $callbackResult instanceof Ray ? $callbackResult : $ray;
+    }
+
+    protected function sendExceptionPayload(Ray $ray): Ray
+    {
+        $exception = array_shift(Ray::$caughtExceptions);
+
+        return $ray->exception($exception);
+    }
+
+    protected function processCallback(Ray $ray, $callback, $rethrow = true): Ray
```

Comment:
This is a rather long function. Could we break this up in smaller functions with a good descriptive name?